Open Judge System/Web/OJS.Web/Areas/Contests/Controllers/CompeteController.cs

```diff
@@ -80,79 +100,89 @@ public void ValidateContest(Contest contest, bool official)
+                throw new HttpException(
+                    (int)HttpStatusCode.NotFound,
+                    Resource.ContestsGeneral.Contest_not_found);
             }
 
             if (official &&
-                !contest.CanBeCompeted &&
-                !(this.User.IsAdmin() ||
-                    contest.Lecturers.Any(c => c.LecturerId == this.UserProfile?.Id) ||
-                    contest.Category.Lecturers.Any(cl => cl.LecturerId == this.UserProfile?.Id)))
+                !this.IsUserAdminOrLecturerInContest(contest) &&
+                !this.participantsBusiness.CanCompeteByContestAndUserId(contest, this.UserProfile.Id))
             {
-                throw new HttpException((int)HttpStatusCode.Forbidden, Resource.ContestsGeneral.Contest_cannot_be_competed);
+                throw new HttpException(
+                    (int)HttpStatusCode.Forbidden,
+                    Resource.ContestsGeneral.Contest_cannot_be_competed);
             }
 
             if (!official && !contest.CanBePracticed)
             {
-                throw new HttpException((int)HttpStatusCode.Forbidden, Resource.ContestsGeneral.Contest_cannot_be_practiced);
+                throw new HttpException(
+                    (int)HttpStatusCode.Forbidden,
+                    Resource.ContestsGeneral.Contest_cannot_be_practiced);
             }
-        }
+        } 
 
         /// <summary>
         /// Displays user compete information: tasks, send source form, ranking, submissions, ranking, etc.
         /// Users only.
         /// </summary>
         [Authorize]
-        public ActionResult Index(int id, bool official)
+        public ActionResult Index(int id, bool official, bool? hasConfirmed)
         {
-            var contest = this.Data.Contests.GetById(id);
-            this.ValidateContest(contest, official);
+            var contest = this.contestsData.GetById(id);
 
-            if (official)
+            try
             {
-                try
-                {
-                    // TODO: check also if ContestType is online
-                    if (!this.IsUserAllowedToCompete(this.UserProfile.Id, id, Settings.ApiKey))
-                    {
-                        this.TempData.AddDangerMessage(Resource.ContestsGeneral.User_is_not_registered_for_exam);
-                        return this.RedirectToAction<HomeController>(c => c.Index(), new { area = string.Empty });
-                    }
-                }
-                catch
-                {
-                    this.TempData.AddDangerMessage(Resource.ContestsGeneral.Contest_cannot_be_competed);
-                    return this.RedirectToAction<HomeController>(c => c.Index(), new { area = string.Empty });
-                }
+                this.ValidateContest(contest, official);
 
-                if (!this.ValidateContestIp(this.Request.UserHostAddress, id))
+                if (official &&
+                    !contest.IsOnline &&
```

Comment:
Why do we exclude Online contests from Ip validation, the fact that they'll probably be opened, does not mean that no IP validation should happen.